File: src/perl/features.pl
###### Class FEATURES -- Features of transactions.
###	$Id$
###
###	A FEATURES object is attached to a transaction when it is
###	first noticed by the Resolver.  It is used to perform lazy
###	evaluation of named features.  The association between names
###	and functions is made using a hash: %FEATURES::computers.  
###
###	The FEATURES object is normally accessed only by a method on
###	TRANSACTION:  transaction->is('feature').  Note that there is
###	no backlink to a transaction from its features; this makes
###	transactions easier to delete by avoiding circularities.
###
###	Features may actually have any value, but the resolver is only
###	interested in truth value.  Some agents, however, make use of
###	feature values.  In particular, 'agent' binds to the name of the
###	agent at which a request is directed.
###	
###	An agent can easily create a transaction and assert features in
###	order to ensure that some other agent will or will not match it.  It
###	is also possible to add a feature and ``recycle'' a transaction.  In
###	this case we will eventually need a feature to prevent unwanted
###	matching by agents that have already seen it.

package FEATURES;

%computers;			# the routines that compute features.

sub new {
    my ($class, $transaction) = @_;

    ## Create a new FEATURES.  We pass the transaction, even though 
    ##    no link is kept, so that we can compute a few features that
    ##    are closely related and that we know are always needed.

    my $self = {'NEVER'=>0};

    ## NEVER is useful for creating agents that never match; these
    ##    have their handlers pushed by the agency.  A null criteria
    ##    list will always match, so there's no need for an ALWAYS.

    bless $self,$class;

    ## Set the transaction's features to $self.
    ##    This is essential in order for initialization to work.

    $transaction->features($self);
    $self->initialize($transaction);
    return $self;
}

### Access to feature values.

sub assert {
    my ($self, $feature, $value) = @_;

    ## Assert a named feature, with an optional value.

    $value = 1 unless defined $value;
    $value = 0 if $value eq '';
    print " $feature=>$value" if $main::debugging;
    $$self{$feature} = $value;
    return $value;
}

sub deny {
    my ($self, $feature) = @_;

    ## Deny a named feature, i.e. assign it a value of 0

    print " $feature=>0" if $main::debugging;
    $$self{$feature} = 0;
    return 0;
}

sub has {
    my ($self, $feature) = @_;

    ## Test for the presence of a named feature

    return defined $$self{$feature};
}

sub test {
    my ($self, $feature, $transaction) = @_;

    ## Test a named feature and return its value.

    my $value = $$self{$feature};
    $value = $self->compute($feature, $transaction) if ! defined $value;
    return $value;
}

sub compute {
    my ($self, $feature, $transaction) = @_;

    ## Compute and assert the value of the given feature.
    ##	 Can be used to recompute features after changes

    my $computer = $computers{$feature};
    print "computing $feature with $computer \n" if $main::debugging;

    return $self->assert($feature, &{$computer}($transaction)) 
	if defined $computer;

    ## Not defined:
    ##	  Undefined features get asserted as a null string, so you can
    ##	  tell the difference between false and undefined, but don't
    ##	  have to recompute things you know aren't there.

    print "No computer for feature $feature\n" if $main::debugging;
    return $self->assert($feature, '');
}

### Initialization:

sub initialize {
    my ($self, $transaction) = @_;

    ## Compute and assert the value of some initial features.
    ##    We do this here because the features are closely related, so
    ##    we can get many assertions out of a small number of requests.

    if ($transaction->is_request) {
	$self->assert('request');
	my $f = $self->compute('agent_request', $transaction);
	$self->assert('proxy_request', ! $f);

	$self->deny('response');
	$self->deny('proxy_response');
	$self->deny('agent_response');
    } else {
	$self->assert('response');
	my $f = $self->compute('agent_response', $transaction);
	$self->assert('proxy_response', ! $f);

	$self->deny('request');
	$self->deny('proxy_request');
	$self->deny('agent_request');
    }
}

sub register {
    my ($feature, $sub) = @_;

    ## Register a subroutine that computes a feature.

    if (defined $computers{$feature}) {
	print "Multiply-defined feature: $feature\n" if ! $main::quiet;
	return;			# === dubious ===
    }
    $computers{$feature} = $sub;
}

############################################################################
###
### Matching.
###
###	The match criteria are a list (not a hash, because order might be
###	significant), of sublists or name=>value pairs.
###
###		[criteria]	ORed, i.e. fail if not matched.
###		x => bool	fail if !test(x) != !b
###		x => \&subr	fail if &subr(test(x)) returns false
###		x => \$var	$var = test(x)

### === maybe 	x => ["op" value]	comparison
###		x => [value] 		eq
###		x => [subr args...] (splice test(x) in as first arg.)

sub matches {
    my($self, $criteria)=@_;

    ## Match against a list of criteria.
    ##	  The criteria are a list because order might be important.
    ##    We index through the list so we don't have to copy it.

    my $i;
    for ($i = 0; $i <= $#$criteria; $i++) {
	my $c = $$criteria[$i];
	if (! ref $c) {
	    local $f = $self->test($c);
	    local $v = $$criteria[++$i];
	    print " $f" if $main::debugging;
	    if (! ref $v) {
		return 0 if !$f != !$v;
	    } elsif (ref($v) eq 'REF') {
		$$v = $f;
	    } elsif (ref($v) eq 'CODE') {
		return 0 if ! &$v($f);
	    } else {
		print "undefined match criterion $c=>$v\n";
	    }
	} elsif (ref($c) eq 'ARRAY') {
	    return 0 if ! self->matches($c);
	} else {
	    print "undefined match criterion $c\n";
	}
    }
    print " matched.\n" if  $main::debugging;
    return 1;
}

############################################################################
###
### Feature Computers.
###	All take a transaction as their argument, and most return a
###	boolean.  Feature computers may use the utility method
###	transaction->assert(name,value) to set additional features. 
###
###	By convention, a feature computer "is_foo" computes a feature
###	named "foo". 

### Default Features: 
###	These are computed by default when a transaction is created;
###	they may have to be recomputed if the transaction is modified.

$computers{response} = \&is_response;
sub is_response{
    my $request=shift;
    return $request->is_response;
}

$computers{request} = \&is_request;
sub is_request{
    my $request=shift;
    return $request->is_request;
}

$computers{agent_response} = \&is_agent_response;
sub is_agent_response{
    my $request=shift; 		return 0 unless $request->is_response;
    my $agent=$request->header('Version');
    return 1 if $agent =~ /^PIA/i;
    return 0;
}

$computers{proxy_request} = \&is_proxy_request;
sub is_proxy_request{
    my $request=shift;    	return 0 unless $request->is_request;
    my $url=$request->url; 	return 0 unless $url;
    
    my $host=$url->host;

    return 0 if ($host =~ /^agency/ || $host eq '');
    return 0 if ($url->port == $main::PIA_PORT && $main::PIA_HOST =~ /^$host/);

    return 1 if $host;
    return 0;
}

$computers{agent_request} = \&is_agent_request;
sub  is_agent_request{
    my $request=shift; 		return 0 unless $request->is_request;
    my $url=$request->url; 	return 0 unless defined $url;
    my $host=$url->host;

    if ($host=~/^agency/ || $host eq ''
	|| $url->port == $main::PIA_PORT && $main::PIA_HOST =~ /^$host/) {
	return 1;
    }
    return 0;
}


### Non-default Features:

$computers{text} = \&is_text;
sub is_text{
    my($request)=shift;
    return $request->content_type() =~ /^text/i;
}

$computers{html} = \&is_html;
sub is_html{
    my($request)=shift;
    return $request->content_type() =~ /^text\/html/i;
}

$computers{image} = \&is_image;
sub is_image{
    my($request)=shift;
    return $request->content_type() =~ /^image/i;
}

$computers{local} = \&is_local;
sub is_local{
    my $request=shift;
    my $url=$request->url; 	return 1 unless $url;
    my $host=$url->host;

    return 1 if ($host =~ /^agency/ || $host eq '');
    return 1 if ($main::PIA_HOST =~ /^$host/);
    return 1 if ($host =~ /localhost/i);
    return 0;
}

$computers{local_source} = \&is_local_source;
sub is_local_source{
    my $request=shift;
    ##need to fill transaction model first
}

$computers{client_is_netscape} = \&client_is_netscape;
sub client_is_netscape{
    my($request)=shift;
    return $request->header()->header('User-Agent') =~ /netscape/i;
}

$computers{file_request} = \&is_file_request;
sub is_file_request{
    my $request=shift;
    my $url=$request->url;
    my $scheme=$url->scheme;
    return $scheme=~/file/i;
}

$computers{interform} = \&is_interform;
sub is_interform{
   my $request=shift;
    my $url=$request->url;
 
   my $path=$url->path;
   return $path=~/\.if$/i;
 
}


### Features with values:

$computers{agent} = \&agent;
sub agent {
    my($request)=shift;
    $request = $request->request if $request->is_response;
    return unless defined $request;
    my $url=$request->url;
    return unless defined $url;
    my $path=$url->path;
    return unless defined $url;
    my $name = ($path =~ m:^/(\w+)/*:i) ? $1 : 'agency';
    return $name;
}


1;

File: src/java/crc/ds/Features.java
// Features.java
// $Id$
// (c) COPYRIGHT Ricoh California Research Center, 1997.

/**
 *	A Features list is attached to a Thing when it is first 
 *	needed.  It is used to perform lazy evaluation of named
 *	features.  The association between names and functions is made
 *	using a hash; a reference to this is returned by $f->computers, 
 *	which returns a reference to a class-specific hash that
 *	associates feature names with functions.
 *
 *	The Features list is normally accessed only by a method on
 *	its parent:  $something->is('feature').  Note that there is
 *	no backlink to the parent from its features; this makes
 *	objects easier to delete by avoiding circularities.
 *
 *	Features may actually have any value, but the matching process
 *	is normally only interested in truth value.  Some agents,
 *	however, make use of feature values.  In particular, 'agent'
 *	binds to the name of the agent at which a request is directed.
 */

package crc.ds;

import java.util.Hashtable;
import java.util.Vector;
import crc.pia.Transaction;
import crc.pia.Agent;
import crc.tf.UnknownNameException;

public class Features{
  /**
   * Attribute index - feature table
   * 
   */
  protected Hashtable featureTable;

  /**
   * Compute and assert the value of some initial features.
   *    We do this here because the features are closely related, so
   *    we can get many assertions out of a small number of requests.  
   */
  protected void initialize( Object parent ){
    
    featureTable.put( "NEVER", new Boolean( false ) );
    
    /*
    * NEVER is useful for creating things that never match.
    *	  A null criteria list will always match, so there's no need
    *	  for an ALWAYS.
    */
  }

  /**
   * Assert a named feature, with a value. 
   */
  public Object assert(String feature, Object v){
    Object value;

    if(v==null){
      value = new Boolean( true );
      featureTable.put( feature, value );
    }else if( v instanceof String && v == "" ){
      value = new Boolean( false );
      featureTable.put( feature, value );
    }else {
        featureTable.put( feature, v );
	value = v;
    }
    return value;
  }

  /**
   * Assert a named feature. 
   */
  public Object assert(String feature){
    Object value = new Boolean( true );

    featureTable.put( feature, value );
    return value;
  }


  /**
   *Deny a named feature, i.e. assign it a value of false
   */
  public boolean deny(String feature){
    featureTable.put( feature, new Boolean( false ) );
    return false;
  }

  /**
   *Test for the presence of a named feature
   */
  public boolean has(String feature){
    return featureTable.containsKey( feature );
  }

  /**
   * Test a named feature and return a boolean.
   */
  public boolean test(String feature, Object parent){
    Object value = featureTable.get( feature );
    
    if (value == null) value = compute(feature, parent);

    return test(value);
  }

  /** Test a value.  
   *	@return false for null, False, and "", true otherwose.
   */
  public static final boolean test(Object value) {
    if (value == null) return false;
    else if (value instanceof Boolean) {
      return ((Boolean)value).booleanValue();
    } else if (value instanceof String) {
      return ! "".equals(value);
    } else {
      return true;
    }
  }

  /**
   *  Return the raw value of the feature.
   */
  public Object feature( String featureName, Object parent ){
    Object val ;

    val = featureTable.get( featureName );
    if ( val == null )
      return compute(featureName, parent); 
    return val;
    
  }


  /**
   * Associate a named feature with an arbitrary value.
   *
   */
  public Object setFeature( String featureName, Object value ){
    return assert( featureName, value );
    
  }


  /**
   * Compute and assert the value of the given feature.
   * Can be used to recompute features after changes
   */
  public Object compute(String feature, Object parent){
    Object val;

    try{
      if( parent instanceof Transaction )
	val = ((Transaction) parent).computeFeature( feature );
      else
	val = ((Agent) parent).computeFeature( feature );
      return setFeature( feature, val );
    }catch(UnknownNameException e){
      return assert(feature, "");
    }
  }


  /**
   * Create a new Features for a Transaction.  Compute a few features that
   * are closely related and that we know are always needed.
   */

  public Features (Transaction parent) {
    featureTable = new Hashtable();
    parent.setFeatures( this );
    initialize( parent );
   }

  public Features() {
    featureTable = new Hashtable();
    initialize(null);
  }

  /**
   * Matching.
   *
   *	The match criteria are a list (not a hash, because order might be
   *	significant), of sublists or name=>value pairs.
   *
   *		[criteria]	ORed, i.e. fail if not matched.
   *		x => bool	fail if !test(x) != !b
   *		x => \&subr	fail if &subr(test(x)) returns false
   *		x => \$var	$var = test(x)
   *
   * === maybe 	x => ["op" value]	comparison
   *		x => [value] 		eq
   *
   *            x => [subr args...] (splice test(x) in as first arg.)
   */
  public boolean matches ( Vector criteria, Object parent ) throws NullPointerException{
    int i;
    if( criteria == null ) throw new NullPointerException("bad criteria.");

    for( i = 0; i < criteria.size(); i++){
      Object o = criteria.elementAt( i );

      if( o instanceof String ){
	String c = (String)o;
	boolean feature = test( c, parent );
	
	Object ov = criteria.elementAt( ++i );
	if( ov instanceof Boolean ){
	  Boolean value = (Boolean) ov;
	  if (!feature != !value.booleanValue())
	    return false;
	}
      }

    }
    return true;
  }


  
}









